sympy/functions/combinatorial/factorials.py

```diff
@@ -866,192 +941,455 @@ def fdiff(self, argindex=1):
-
-        k = self.args[1]
-        if (n-k).is_Integer:
+        from sympy import Product
+        n, k = self.args
+        # see if re-evaluation evaluates
+        e = self.func(n, k)
+        if e.func != self.func:
+            return e
+
+        # see if there is a non-negative k or n - k
+        if k.is_number and int_like(k) and k.is_nonnegative:
+            pass
+        elif (n - k).is_number and int_like(n - k) and (n - k).is_nonnegative:
             k = n - k
-
-        if k.is_Integer:
-            if k.is_zero:
-                return S.One
-            elif k.is_negative:
-                return S.Zero
-            else:
-                n, result = self.args[0], 1
-                for i in range(1, k + 1):
-                    result *= n - k + i
-                    result /= i
-                return result
         else:
-            return binomial(*self.args)
+            return self
+        i = Dummy('never_seen')
+        return (Product(n - k + i, (i, 1, k))/factorial(k)).doit()
+
+    def _eval_rewrite_as_Piecewise(self, n, k, **kwargs):
+        from sympy import gamma, Abs, Lt
+        isint = lambda x: Eq(x, floor(x))
+        nneg = lambda x: Or(Eq(x, 0), Eq(x/Abs(x), 1))
+        for i in igen():
+            if not n.has(i) and not k.has(i):
+                break
+        return Piecewise(
+            (factorial(n)/factorial(k)/factorial(n - k),
+                And(isint(n), isint(k),
+                nneg(n), nneg(k), nneg(n - k))),
+            (ff(n, k)/factorial(k),
+                And(isint(k), nneg(k))),
+            (ff(n, n - k)/factorial(n - k),
+                And(isint(n - k), nneg(n - k))),
+            (0, And(*[i for x in (n, k, n - k)
+                for i in [isint(x), Lt(x, 0)]])),
+            (gamma(n + 1)/gamma(k + 1)/gamma(n - k + 1),
+                True))
 
     def _eval_rewrite_as_factorial(self, n, k, **kwargs):
-        return factorial(n)/(factorial(k)*factorial(n - k))
+        from sympy import Pow
+        isint = lambda x: x.is_integer is True
+        nneg = lambda x: x.is_nonnegative is True
+        neg = lambda x: x.is_negative is True
+        if And(isint(n), isint(k)):
+            if And(neg(n), nneg(k)):
+                return(Pow(-1, k)*factorial(k - n - 1)/(factorial(k)*factorial(-n - 1)))
+            if And(neg(n), neg(k)):
+                return (Pow(-1, n - k)*factorial(-k - 1)/(
+                    factorial(n - k)*factorial(-n - 1)))
+            if And(nneg(n), neg(k)):
+                return S.Zero
+            if And(nneg(n), nneg(k)):
+                return factorial(n)/(factorial(k)*factorial(n - k))
 
     def _eval_rewrite_as_gamma(self, n, k, **kwargs):
-        from sympy import gamma
-        return gamma(n + 1)/(gamma(k + 1)*gamma(n - k + 1))
+        from sympy import gamma, Pow
+        ki = k.is_integer
+        nki = (n - k).is_integer
+        if ki is None:
+            return
+        if nki is None:
+            return
```

Comment:
`fuzzy_or` weeds out `None` right ?
we could rewrite in terms of that

Reply:
I like to see coverage. Being explicit like this verifies that both cases have been considered.

Reply:
Hmm...taken to it's conclusion, no routine like this would ever use `or` or `and`. So I guess I could go with `if None not in (ki, nki): return`.